test/fail/objpat-type-fields.mo

```diff
@@ -0,0 +1,74 @@
+import { type T; type Result } "objpat-type-fields/lib";
+
+func _result() {
+  type MyNat = T;
+
+  let _t : MyNat = 1;
+  let _r1 : Result<Nat, Text> = #ok(1);
+  let _r2 : Result<Nat> = #ok(3); // Wrong number of type parameters
+};
+
+func _bounds() {
+  type Bounded<T <: Result<Int, Text>> = T;
+  type BFine = Bounded<Result<Nat, Text>>;
+  type BWrong = Bounded<Result<Float, Text>>;
+};
+
+type L1 = module { type N = Nat };
+type R1 = module { type N = Text };
+type V1 = { #left : L1; #right : R1 };
+
+func _f1(x : V1) {
+  // Fine
+  let (#left{ type N }) = x else {
+    return ()
+  };
+
+  let _n : N = 10;
+};
+
+func _f1_1(x : V1) {
+  // Error with mismatched types
+  let (#left{ type N } or #right { type N }) = x else {
```

Comment:
Interesting and confusing 😄 In the original issue I thought we just wanted unqualified imports only in `import`s on the toplevel.
This is complicating the compiler code, right? You need to traverse the pattern to add potential types to the scope. Complex logic that will probably be unused by most users, but would need to be maintained.

Reply:
Ok since imports are desugared to lets I guess it's the easiest to allow type patterns in every object pattern